torch/nn/functional.py

```diff
@@ -3145,73 +3158,94 @@ def multi_head_attention_forward(query,                  # type: Tensor
     scaling = float(head_dim) ** -0.5
 
-    if qkv_same:
-        # self-attention
-        q, k, v = linear(query, in_proj_weight, in_proj_bias).chunk(3, dim=-1)
-
-    elif kv_same:
-        # encoder-decoder attention
-        _b = in_proj_bias
-        _start = 0
-        _end = embed_dim
-        _w = in_proj_weight[_start:_end, :]
-        if _b is not None:
-            _b = _b[_start:_end]
-        q = linear(query, _w, _b)
-
-        if key is None:
-            assert value is None
-            k = None
-            v = None
+    if use_chunk_proj_weight:
+        if qkv_same:
+            # self-attention
+            q, k, v = linear(query, in_proj_weight, in_proj_bias).chunk(3, dim=-1)
+
+        elif kv_same:
+            # encoder-decoder attention
+            _b = in_proj_bias
+            _start = 0
+            _end = embed_dim
+            _w = in_proj_weight[_start:_end, :]
+            if _b is not None:
+                _b = _b[_start:_end]
+            q = linear(query, _w, _b)
+
+            if key is None:
+                assert value is None
+                k = None
+                v = None
+            else:
+
+                _b = in_proj_bias
+                _start = embed_dim
+                _end = None
+                _w = in_proj_weight[_start:, :]
+                if _b is not None:
+                    _b = _b[_start:]
+                k, v = linear(key, _w, _b).chunk(2, dim=-1)
+
         else:
+            _b = in_proj_bias
+            _start = 0
+            _end = embed_dim
+            _w = in_proj_weight[_start:_end, :]
+            if _b is not None:
+                _b = _b[_start:_end]
+            q = linear(query, _w, _b)
 
             _b = in_proj_bias
             _start = embed_dim
+            _end = embed_dim * 2
+            _w = in_proj_weight[_start:_end, :]
+            if _b is not None:
+                _b = _b[_start:_end]
+            k = linear(key, _w, _b)
+
+            _b = in_proj_bias
+            _start = embed_dim * 2
             _end = None
             _w = in_proj_weight[_start:, :]
             if _b is not None:
                 _b = _b[_start:]
-            k, v = linear(key, _w, _b).chunk(2, dim=-1)
-
+            v = linear(value, _w, _b)
     else:
-        _b = in_proj_bias
-        _start = 0
-        _end = embed_dim
-        _w = in_proj_weight[_start:_end, :]
-        if _b is not None:
-            _b = _b[_start:_end]
-        q = linear(query, _w, _b)
-
-        _b = in_proj_bias
-        _start = embed_dim
-        _end = embed_dim * 2
-        _w = in_proj_weight[_start:_end, :]
-        if _b is not None:
-            _b = _b[_start:_end]
-        k = linear(key, _w, _b)
-
-        _b = in_proj_bias
-        _start = embed_dim * 2
-        _end = None
-        _w = in_proj_weight[_start:, :]
-        if _b is not None:
-            _b = _b[_start:]
-        v = linear(value, _w, _b)
+        assert q_proj_weight is not None
```

Comment:
Add a message to be printed if the assert fails or catch it higher up.

Reply:
the jit team instructed me to convert optional tensor q_proj_weight to non-optional tensor q_proj_weight_non_opt. The casting will throw error message if q_proj_weight is None.